### PR TITLE
SLCORE-29 Check language plugins compatibility

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/WsHelperImpl.java [hunk 1]
```diff
@@ -39,12 +43,12 @@ public ValidationResult validateConnection(ServerConfiguration serverConfig) {
...
-      if (result.success()) {
-        return new AuthenticationChecker(client).validateCredentials();
-      } else {
-        return result;
-      }
+      PluginVersionChecker pluginsChecker = new PluginVersionChecker(client);
+      ServerInfos serverInfo = serverChecker.checkVersionAndStatus();
+      pluginsChecker.checkPlugins(serverInfo.getVersion());
+      return new AuthenticationChecker(client).validateCredentials();
+    } catch (UnsupportedServerException e) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/WsHelperImpl.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/GlobalUpdateExecutor.java [hunk 2]
```diff
@@ -58,39 +61,49 @@ public GlobalUpdateExecutor(StorageManager storageManager, SonarLintWsClient wsC
...
+      ProtobufUtil.writeToFile(updateStatus, temp.resolve(StorageManager.UPDATE_STATUS_PB));
+
+      Path dest = storageManager.getGlobalStorageRoot();
+      FileUtils.deleteDirectory(dest);
+      FileUtils.forceMkDirs(dest.getParent());
+      FileUtils.moveDir(temp, dest);
+    } catch (RuntimeException e) {
+      try {
+        FileUtils.deleteDirectory(temp);
+      } catch (RuntimeException ignore) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/GlobalUpdateExecutor.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/validate/PluginVersionChecker.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        .append(p.getVersion())
+        .append(", minimum: ")
+        .append(props.getProperty(p.getKey()))
+        .append(")");
+    }
+
+    return builder.toString();
+  }
+
+  private boolean validate(String minVersion, String version) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/validate/PluginVersionChecker.java]: MAJOR Make "validate" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/validate/PluginVersionChecker.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  public void checkPlugins(String serverVersion) {
+    ValidationResult result = validatePlugins(serverVersion);
+    if (!result.success()) {
+      throw new UnsupportedServerException(result.message());
+    }
+  }
+
+  private String buildFailMessage(Properties props, List<InstalledPlugin> failingPlugins) {
```
Comment (SonarTech) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/validate/PluginVersionChecker.java]: MAJOR Make "buildFailMessage" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 4 issues
* CRITICAL 2 critical
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 2 issues
* CRITICAL 2 critical

Watch the comments in this conversation to review them.

File: core/src/main/java/org/sonarsource/sonarlint/core/container/connected/validate/PluginVersionChecker.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  public void checkPlugins(String serverVersion) {
+    ValidationResult result = validatePlugins(serverVersion);
+    if (!result.success()) {
+      throw new UnsupportedServerException(result.message());
+    }
+  }
+
+  private static String buildFailMessage(Properties props, List<InstalledPlugin> failingPlugins) {
```
Comment (henryju) [inline core/src/main/java/org/sonarsource/sonarlint/core/container/connected/validate/PluginVersionChecker.java]: maybe rename "props" by something more meaningful like "minimalPluginVersions"

File: core/src/test/resources/validate/installed_plugins.json [hunk 6]
```diff
@@ -0,0 +1 @@
... a SVG image (so called badge.)","version":"1.2.1","category":"Integration","license":"The GNU Lesser General Public License, Version 3.0","organizationName":"QualInsight","organizationUrl":"http://www.qualinsight.com/","homepageUrl":"https://github.com/QualInsight/qualinsight-plugins-sonarqube-badges/"},{"key":"scmsvn","name":"SVN","description":"SVN SCM Provider.","version":"1.3","category":"Integration","license":"GNU LGPL 3","organizationName":"SonarSource","organizationUrl":"http://www.sonarsource.com","homepageUrl":"http://redirect.sonarsource.com/plugins/scmsvn.html","issueTrackerUrl":"https://jira.sonarsource.com/browse/SONARSCSVN","implementationBuild":"73d9c8a5764e95388328bbaeb911f16fc925a448"},{"key":"timeline","name":"Timeline","description":"Advanced time machine chart (requires an internet access).","version":"1.5","category":"Visualization/Reporting","license":"GNU LGPL 3","organizationName":"SonarSource","organizationUrl":"http://www.sonarsource.com","homepageUrl":"http://docs.codehaus.org/display/SONAR/Timeline+Plugin","issueTrackerUrl":"http://jira.codehaus.org/browse/SONARPLUGINS/component/14068","implementationBuild":"a9cae1328fd455a128b5d7d603381f47398c6e2a"}]}
```
Comment (henryju) [inline core/src/test/resources/validate/installed_plugins.json]: Could you format the file to make it easier to change later?

File: core/src/test/resources/validate/installed_plugins_fail.json [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    "category":"Languages",
+    "license":"GNU LGPL 3",
+    "organizationName":"SonarSource",
+    "organizationUrl":"http://www.sonarsource.com",
+    "homepageUrl":"http://redirect.sonarsource.com/plugins/java.html",
+    "issueTrackerUrl":"https://jira.sonarsource.com/browse/SONARJAVA",
+    "implementationBuild":"af955c8affd056240393371c34c2a78bea69a114"
+  },
+  {
+    "key":"javadebuggingrules",
```
Comment (henryju) [inline core/src/test/resources/validate/installed_plugins_fail.json]: Maybe cleanup this file to not expose our internal stuff?

**Comment by henryju** [comment]
Few comments then LGTM

**Comment by dbmeneses** [comment]
feedback applied

**Pull request merged**